Rakefile

```diff
@@ -172,14 +172,32 @@ END_HTML_TAIL
 end
 
 require 'rake/extensiontask'
-require 'rspec/core/rake_task'
-RSpec::Core::RakeTask.new(:spec)
 
 Rake::ExtensionTask.new('RMagick2') do |ext|
   ext.ext_dir = 'ext/RMagick'
 end
 
-task spec: :compile
+task spec: :compile do
+  require 'timeout'
+
+  pid = nil
+  try_count = 0
+  begin
+    puts "\e[33mStart running spec\e[0m"
+
+    try_count += 1
+    Timeout.timeout(180) do
+      pid = Process.spawn("bundle exec rspec")
+      Process.wait(pid)
+    end
+  rescue Timeout::Error
+    puts "\e[31mTimeout error\e[0m"
+
+    # As safe retry, it need to kill a process which is running "bundle exec rspec" because it may not be dead.
+    Process.kill(:SIGKILL, pid)
+    retry if try_count < 3
+  end
```

Comment:
@Watson1978 I like the idea of a timeout, but will it work to implement it in an `around` block instead?

```rb
RSpec.configure do |config|
  config.around do |example|
    Timeout.timeout(180) { example.run }
  end
end
```

I'd rather not retry *every* test, since it might be good to know if there are specific tests that are timing out. We could manually flag tests, maybe something like:

```rb
MAX_SPEC_RETRIES = 3
SPEC_TIMEOUT = 180 # seconds

RSpec.configure do |config|
  config.around do |example|
    retry_count = 0
    Timeout.timeout(SPEC_TIMEOUT) { example.run }
  rescue Timeout::Error
    raise unless example.metadata[:retry] && retry_count < MAX_SPEC_RETRIES

    retry_count += 1
    Timeout.timeout(SPEC_TIMEOUT) { example.run }
  end
end

# some test that we know times out sometimes
it 'sometimes takes too long', :retry do
  ...
end
```

Reply:
Hmm, looks like the error is happening before the test, so my approach might not solve the issue.